---
title: "DNA-nn"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


### Read the dataframe from the files

## Including Plots
## plots above ratio in changes between the false and true values between carb and toby vals

You can also embed plots, for example:
#this segment is used remove NA values from the data 
and provide a summary about the two files being read
```{r}
pseudo<-read.csv("/home/aryan/Desktop/DNA-nn/responses-pseudo.csv")

summary(pseudo)
good<-complete.cases(pseudo)
pseudo<-pseudo[good,]
counter<-c(pseudo['carb'][])

staph<-read.csv("/home/aryan/Desktop/DNA-nn/responses-staph.csv")
good<-complete.cases(staph)
staph<-staph[good,]
summary(staph)

```


```{r}
library(tidyverse)
library(tidyr)
library(dplyr)
library(ggplot2)
good<-complete.cases(pseudo)
pseudo<-pseudo[good,]
summary(pseudo)
```

#The following code segment plots the number of TRUE/FALSE for each value we are considering

```{r}
ggplot(pseudo, aes(factor(carb))) + geom_bar(width = 0.5, fill="light blue", color= "red") + ggtitle("(A) Count of responses to carbenicillin (in pseudomonas)")
ggplot(pseudo, aes(factor(toby))) + geom_bar(width = 0.5, fill="light blue", color= "red") + ggtitle("(B) Count of responses to tobramycin (in pseudomonas)")
ggplot(staph, aes(factor(resp))) + geom_bar(width = 0.5, fill="light pink", color= "black") + ggtitle("(C) Count of responses (in Staphylococcus)")

```

#gives the distance matrix and phylogenetic tree for the given .fasta file.
do recommend reading the source file pages for examples and how to install the packages/dependencies. Does not install
in one try
if it does, let me know because I have Windows

```{r}
library(adegenet)
dna <- fasta2DNAbin(file="/home/aryan/Desktop/DNA-nn/concatenated.fasta")
library(ape)
D <- dist.dna(dna, model = "TN93")
length(D)
temp <- as.data.frame(as.matrix(D))
table.paint(temp, cleg=0, clabel.row=.5, clabel.col=.5) #darker shades of gray mean a larger distance # you can also make cool color plots but they're much more complicated because they use the image() function
tre3 <- as.phylo(hclust(D,method="average"))
plot(tre3, cex=.5)
title("UPGMA tree")
```


```{r}
#library(adegenet)
#gen_light <- fasta2genlight(file="/home/aryan/Desktop/DNA-nn/concatenated.fasta")
#grp <- find.clusters(gen_light, max.n=50, n.pca=200, scale=FALSE)
tre <- nj(D)
class(tre)
tre <- ladderize(tre)
tre
plot(tre, cex = 0.6)
title("A Simple NJ Tree")
```


```{r}
h_cluster <- hclust(D, method = "average", members = NULL) # method = average is used for UPGMA, members can be equal to NULL or a vector with a length of size D
plot(h_cluster, cex = 0.6)
```



#CLUSTERING USING THE TRUE FALSE VALUES
#we can use the total area method 
#Do consider that 
```{r}
plot(tre, show.tip=FALSE) # gets rid of the labels on the end, refer to the first tree depicted above
title("Unrooted NJ tree")
myPal <- colorRampPalette(c("red","blue"))
library(dplyr)
pseudo <- pseudo %>% mutate(carb2Num = as.numeric(carb))
tiplabels(pseudo$carb2Num, bg=num2col(pseudo$carb2Num, col.pal=myPal), cex=.5) #we use the responses dataset to get the TRUE/FALSE values
temp <- pretty(2:3, 2)
legend("bottomleft", fill=num2col(temp, col.pal=myPal), leg=temp, ncol=2)
```



```{r}
library(seqinr)
#staph <- read.alignment(file="/home/aryan/Desktop/DNA-nn/core_gene_alignment-narsa.aln", format="fasta")
library(ape)
library(adegenet)
staphDNAbin<-fasta2DNAbin(file="/home/aryan/Desktop/DNA-nn/core_gene_alignment-narsa.aln")
```


```{r}
D <- dist.dna(staphDNAbin, model = "TN93")
length(D)
temp <- as.data.frame(as.matrix(D))
table.paint(temp, cleg=0, clabel.row=.5, clabel.col=.5) #darker shades of gray mean a larger distance # you can also make cool color plots but they're much more complicated because they use the image() function
tre3 <- as.phylo(hclust(D,method="average"))
plot(tre3, cex=.5)
title("UPGMA tree")

#used for Neighbor joining tree. 
tre <- nj(D)
class(tre)
tre <- ladderize(tre)
tre
plot(tre, cex = 0.6)
title("A Simple NJ Tree")
h_cluster <- hclust(D, method = "average", members = NULL) # method = average is used for UPGMA, members can be equal to NULL or a vector with a length of size D
plot(h_cluster, cex = 0.6)
plot(tre, show.tip=FALSE) # gets rid of the labels on the end, refer to the first tree depicted above
title("Unrooted NJ tree")
myPal <- colorRampPalette(c("red","blue"))

#library(dplyr)
#staph <- staph %>% mutate(resp2Num = as.numeric(resp))
#tiplabels(staph$resp2num, bg=num2col(staph$resp2num, col.pal=myPal), cex=.5) #we use the responses dataset to get the TRUE/FALSE values
#temp <- pretty(2:3, 2)
#legend("bottomleft", fill=num2col(temp, col.pal=myPal), leg=temp, ncol=2)
```
#```{r}
library(mice)
md.pattern(phyto)
phy<-as.character(phyto)
mice_plot <- aggr(phy, col=c('navyblue','yellow'),
numbers=TRUE, sortVars=TRUE,
labels=names(iris.mis), cex.axis=.7,
gap=3, ylab=c("Missing data","Pattern"))
dna.mis <- prodNA(phy, noNA = 0.1)
```

